plenum/test/view_change/test_view_change_with_instance_change_lost_due_to_restarts.py

```diff
@@ -0,0 +1,42 @@
+import pytest
+
+from plenum.test.helper import freshness, waitForViewChange
+from plenum.test.node_request.helper import sdk_ensure_pool_functional
+from plenum.test.restart.helper import restart_nodes
+from plenum.test.test_node import ensureElectionsDone
+
+FRESHNESS_TIMEOUT = 20
+
+
+@pytest.fixture(scope="module")
+def tconf(tconf):
+    with freshness(tconf, enabled=True, timeout=FRESHNESS_TIMEOUT):
+        yield tconf
+
+
+def test_view_change_with_instance_change_lost_due_to_restarts(looper, txnPoolNodeSet,
+                                                               sdk_pool_handle,
+                                                               sdk_wallet_client,
+                                                               tconf, tdir, allPluginsPath):
+    """
+    1. some_nodes (Beta and Gamma) send InstanceChange for all nodes.
+    2. Restart other_nodes (Gamma and Delta)
+    3. last_node (Delta) send InstanceChange for all nodes.
+    4. Ensure elections done and pool is functional
```

Comment:
If I understand correctly, in this test we have the following situation after step3:
- Alpha (Primary) has 3 IC and started VC
- Beta has 3 IC and started VC
- Gamma has 1 IC 
- Delta has 1 IC
Since Primary stopped ordering, freshness checks lead to start of a VC on all nodes.
So, it makes sense to call the test like `test_vc_finished_when_less_than_quorum_started_including_primary`.

But what if a different combination of nodes restarted, so that Primary doesn't have a quorum of IC?
For example, 
- Alpha and Beta send IC
- Beta is restarted (wait until it's restarted before go to the next step)
- Alpha is restarted (wait until it's restarted before go to the next step)
- Gamma is restarted (wait until it's restarted before go to the next step)
- Delta sends IC
= >
Alpha - 1 IC 
Beta - 1 IC 
Gamma - 1 IC
Delta - 3 IC => starts VC.
As a result, Delta will be in infinite VC.

I think the freshness check can not solve the situation above, so persistence of IC will be needed.

Reply:
Another test: all nodes start VC, but 1/2 of the nodes restart during the VC. 
Sub-tests: 
- 1/2 nodes are restarted immediately
- 1/2 nodes are restarted in > than INSTANCE_CHANGE_TIMEOUT
- 1/2 nodes are restarted in > than 3*INSTANCE_CHANGE_TIMEOUT

Reply:
Nice to have:
Maybe we should implement something like unit/property-based tests for checking that we can recover from any state.
Take ViewChange classes, put then into different view, and sent IC and VC messages to it in different combinations. It's expected that all ViewChanger classes should eventually come to the same view.
We can do it with our new property-based model as well. 
But if we could do it with a real View Changer class, we would have more confidence.

Reply:
Actually I tried to implement test where only minority of nodes enter view change, however there are some problems with this:
- it is possible to do so on a 7 nodes pool, however since master primary also enters view change and ordering is stopped this case is no different from already implemented
- preventing primary from collecting instance change quorum by restarting it will lead to primary disconnection events and view change will happen anyways
- fun fact - in INDY-1903 primary was not restarted, but it didn't receive instance change messages during network outages, that's why it didn't enter view change. However it should have received them due to queueing (and I fail to see how to reproduce it correctly in tests), so probably there was some yet another bug?

Reply:
Implementing property based tests for view changer would be extremely helpful, however view changer does depend a lot on node and it uses real communication layer, not simulated one. There are multiple implementation options, but they all have drawbacks:
- just implement full-blown integration property based test - this can be done pretty quickly, however such test would take ages to run, and I have concerns about reproducibility of its results
- use monkeypatching and mocking extensively to isolate view changer without touching it too much - this can be done in moderate amount of time, however I'm afraid such tests would be very fragile as they will depend a lot on implementation details
- refactor view changer so that it can be easily used in isolation - this is the cleanest solution, and it can be used as a foundation for similar changes in other parts of code, but the question is if we are okay with amount of work it would take